crates/metadata-store/src/omnipaxos/store.rs

```diff
@@ -0,0 +1,1232 @@
+                                break Provisioned::Passive(self.become_passive())
+                            }
+                        }
+                    }
+                }
+            }
+        };
+
+        TaskCenter::spawn_unmanaged(TaskKind::Background, "provision-responder", async move {
+            while let Some(request) = provision_rx.recv().await {
+                let _ = request.result_tx.send(Ok(false));
+            }
+        })?;
+
+        Ok(result)
+    }
+
+    fn initialize_storage(
+        &mut self,
+        nodes_configuration: NodesConfiguration,
+    ) -> Result<OmniPaxosConfiguration, InitializeMetadataStoreError> {
+        let (omni_paxos_configuration, nodes_configuration) =
+            self.derive_initial_configuration(nodes_configuration)?;
+
+        debug!("Initialize storage with nodes configuration: {nodes_configuration:?}");
+
+        let value = serialize_value(&nodes_configuration)?;
+
+        Self::prepare_storage(
+            &mut self.rocksdb_storage,
+            &omni_paxos_configuration,
+            vec![Request {
+                request_id: Ulid::new(),
+                kind: RequestKind::Put {
+                    precondition: Precondition::None,
+                    key: NODES_CONFIG_KEY.clone(),
+                    value,
+                },
+            }],
+        )?;
+
+        Ok(omni_paxos_configuration)
+    }
+
+    fn prepare_storage(
+        rocksdb_storage: &mut RocksDbStorage<Request>,
+        omni_paxos_configuration: &OmniPaxosConfiguration,
+        log_prefix: Vec<Request>,
+    ) -> Result<(), StorageError> {
+        // we assume that log_prefix is complete, this will change once we support snapshots
+        rocksdb_storage.batch_set_configuration(omni_paxos_configuration)?;
+        rocksdb_storage.batch_append_on_prefix(0, log_prefix)?;
+        rocksdb_storage.commit_batch()?;
+
+        Ok(())
+    }
+
+    fn derive_initial_configuration(
+        &self,
+        mut nodes_configuration: NodesConfiguration,
+    ) -> Result<(OmniPaxosConfiguration, NodesConfiguration), InvalidConfiguration> {
+        let configuration = Configuration::pinned();
+
+        let restate_node_id = if let Some(node_config) =
+            nodes_configuration.find_node_by_name(configuration.common.node_name())
+        {
+            if let Some(force_node_id) = configuration.common.force_node_id {
+                if force_node_id != node_config.current_generation.as_plain() {
+                    return Err(InvalidConfiguration(format!(
+                        "nodes configuration has wrong plain node id; expected: {}, actual: {}",
+                        force_node_id,
+                        node_config.current_generation.as_plain()
+                    )));
+                }
+            }
+
+            node_config.current_generation.as_plain()
+        } else {
```

Comment:
This case is needed if another node received the provision signal (e.g. a non metadata store node) and tries to provision the metadata store with a  `NodesConfiguration` that does not contain the metadata store node.